extensions/oidc-client/deployment/src/main/java/io/quarkus/oidc/client/deployment/OidcClientBuildStep.java

```diff
@@ -49,31 +70,153 @@ void runtimeInitializeTokenHelper(BuildProducer<RuntimeInitializedClassBuildItem
+    }
+
+    private SyntheticBeanBuildItem syntheticNamedOidcClientBeanFor(String clientName, OidcClientRecorder recorder,
+            OidcClients clients) {
+        return SyntheticBeanBuildItem.configure(OidcClient.class).unremovable()
+                .types(OidcClient.class)
+                .supplier(recorder.createOidcClientBean(clients, clientName))
+                .scope(Singleton.class)
+                .addQualifier().annotation(NamedOidcClient.class).addValue("value", clientName).done()
+                .setRuntimeInit()
+                .destroyer(BeanDestroyer.CloseableDestroyer.class)
                 .done();
-        return Arrays.asList(oidcClientBuildItem, oidcClientsBuildItem);
+    }
+
+    @BuildStep(onlyIf = IsEnabled.class)
+    public void createNonDefaultTokensProducers(
+            BuildProducer<GeneratedBeanBuildItem> generatedBean,
+            OidcClientNamesBuildItem oidcClientNames) {
+
+        ClassOutput classOutput = new GeneratedBeanGizmoAdaptor(generatedBean);
+
+        String targetPackage = DotNames
+                .internalPackageNameWithTrailingSlash(DotName.createSimple(TokensProducer.class.getName()));
+
+        for (String oidcClientName : oidcClientNames.oidcClientNames()) {
+            createNamedTokensProducerFor(classOutput, targetPackage, oidcClientName);
+        }
+    }
+
+    /**
+     * Create a Tokens producer class like follows:
+     * 
+     * <pre>
+     * &#64;Singleton
+     * public class TokensProducer_oidcClientName extends AbstractTokensProducer {
+     *     &#64;Produces
+     *     &#64;NamedOidcClient("oidcClientName")
+     *     &#64;RequestScoped
+     *     public Tokens produceTokens() {
+     *         return getTokens().await().indefinitely();
+     *     }
+     * 
+     *     &#64;Override
+     *     protected Optional<String> clientId() {
+     *         return Optional.of("oidcClientName");
+     *     }
+     * }
+     * </pre>
+     */
+    private String createNamedTokensProducerFor(ClassOutput classOutput, String targetPackage, String oidcClientName) {
+        String generatedName = targetPackage + "TokensProducer_" + sanitize(oidcClientName);
+
+        try (ClassCreator tokensProducer = ClassCreator.builder().classOutput(classOutput).className(generatedName)
+                .superClass(AbstractTokensProducer.class)
+                .build()) {
+            tokensProducer.addAnnotation(DotNames.SINGLETON.toString());
+
+            try (MethodCreator produceMethod = tokensProducer.getMethodCreator("produceTokens", Tokens.class)) {
+                produceMethod.setModifiers(Modifier.PUBLIC);
+
+                produceMethod.addAnnotation(DotNames.PRODUCES.toString());
+                produceMethod.addAnnotation(NamedOidcClient.class.getName()).addValue("value", oidcClientName);
+                produceMethod.addAnnotation(RequestScoped.class.getName());
+
+                ResultHandle tokensUni = produceMethod.invokeVirtualMethod(
```

Comment:
@fwippe I'm not sure this `Uni` code should be generated. How about pushing the code from `OidcClientFilter` which does the same to `AbstractTokenProducer` for it to be reused ?

Reply:
Good suggestion! I pulled the common `getTokens().await().indefinitely()` code up to `AbstractTokensProducer`.